File: clients/dart/lib/caught_up.dart
import 'dart:async';

import 'src/protocol_sync_loop.dart';

final RegExp _canonicalCheckpoint = RegExp(r'^(?:0|[1-9]\d*)$');

class AuthoritativeCheckpointTarget {
  final int protocolVersion;
  final String checkpoint;
  final String? generation;

  const AuthoritativeCheckpointTarget({
    this.protocolVersion = 1,
    required this.checkpoint,
    this.generation,
  });

  factory AuthoritativeCheckpointTarget.fromJson(Map<String, dynamic> json) {
    return AuthoritativeCheckpointTarget(
      protocolVersion: json['protocolVersion'] as int? ?? 0,
      checkpoint: json['checkpoint'] as String? ?? '',
      generation: json['generation'] as String?,
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        'protocolVersion': protocolVersion,
        'checkpoint': checkpoint,
        if (generation != null) 'generation': generation,
      };
}

abstract interface class AuthoritativeCheckpointRequester {
  Future<AuthoritativeCheckpointTarget> requestCheckpoint(
    CaughtUpCancellationToken cancellation,
  );
}

class CaughtUpCancellationToken {
  bool _cancelled = false;
  final Completer<void> _cancelledCompleter = Completer<void>();

  bool get isCancelled => _cancelled;
  Future<void> get whenCancelled => _cancelledCompleter.future;

  void cancel() {
    if (_cancelled) return;
    _cancelled = true;
    _cancelledCompleter.complete();
  }
}

enum CaughtUpBarrierErrorCode {
  invalidTarget,
  invalidLocalCheckpoint,
  invalidated,
  timeout,
  cancelled,
  offline,
}

class CaughtUpBarrierException implements Exception {
  final CaughtUpBarrierErrorCode code;
  final String message;
  final String? targetCheckpoint;
  final String? localCheckpoint;

  const CaughtUpBarrierException(
    this.code,
    this.message, {
    this.targetCheckpoint,
    this.localCheckpoint,
  });

  @override
  String toString() => 'CaughtUpBarrierException($code): $message';
}

class CaughtUpResult {
  final String targetCheckpoint;
  final String checkpoint;
  final String? generation;
  final Duration elapsed;
  final int cycles;
  final bool alreadyCaughtUp;

  const CaughtUpResult({
    required this.targetCheckpoint,
    required this.checkpoint,
    required this.generation,
    required this.elapsed,
    required this.cycles,
    required this.alreadyCaughtUp,
  });
}

bool checkpointReached(String local, String target) {
  _assertCheckpoint(local, local: true);
  _assertCheckpoint(target, local: false);
  return BigInt.parse(local) >= BigInt.parse(target);
}

void validateAuthoritativeCheckpointTarget(
  AuthoritativeCheckpointTarget target, {
  String? expectedGeneration,
}) {
  if (target.protocolVersion != 1) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.invalidTarget,
      'authoritative checkpoint target has unsupported protocolVersion',
      targetCheckpoint: target.checkpoint,
    );
  }
  _assertCheckpoint(target.checkpoint, local: false);
  if (target.generation != null && target.generation!.isEmpty) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.invalidTarget,
      'authoritative checkpoint generation must be non-empty when present',
      targetCheckpoint: target.checkpoint,
    );
  }
  if (expectedGeneration != null && target.generation != expectedGeneration) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.invalidated,
      'checkpoint generation does not match the expected generation',
      targetCheckpoint: target.checkpoint,
    );
  }
}

void _assertCheckpoint(String checkpoint, {required bool local}) {
  if (_canonicalCheckpoint.hasMatch(checkpoint)) return;
  throw CaughtUpBarrierException(
    local
        ? CaughtUpBarrierErrorCode.invalidLocalCheckpoint
        : CaughtUpBarrierErrorCode.invalidTarget,
    '${local ? 'local' : 'target'} checkpoint must be a canonical unsigned decimal string',
    targetCheckpoint: local ? null : checkpoint,
    localCheckpoint: local ? checkpoint : null,
  );
}

Future<T> _boundedForCaller<T>(
  Future<T> operation, {
  required Duration remaining,
  required String target,
  String? local,
  CaughtUpCancellationToken? cancellation,
}) async {
  if (cancellation?.isCancelled ?? false) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.cancelled,
      'caught-up wait was cancelled',
      targetCheckpoint: target,
      localCheckpoint: local,
    );
  }
  if (remaining <= Duration.zero) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.timeout,
      'caught-up wait timed out before the durable checkpoint reached the target',
      targetCheckpoint: target,
      localCheckpoint: local,
    );
  }

  final timeout = Future<T>.delayed(remaining, () {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.timeout,
      'caught-up wait timed out before the durable checkpoint reached the target',
      targetCheckpoint: target,
      localCheckpoint: local,
    );
  });
  if (cancellation == null) {
    return Future.any<T>(<Future<T>>[operation, timeout]);
  }
  final cancelled = cancellation.whenCancelled.then<T>((_) {
    throw CaughtUpBarrierException(
      CaughtUpBarrierErrorCode.cancelled,
      'caught-up wait was cancelled',
      targetCheckpoint: target,
      localCheckpoint: local,
    );
  });
  // Losing this race never cancels `operation`. In particular, a caller timing
  // out must not stop ProtocolSyncLoop.syncNow(), which may be shared by other
  // waiters or background synchronization.
  return Future.any<T>(<Future<T>>[operation, timeout, cancelled]);
}

Future<CaughtUpResult> awaitCaughtUp(
  ProtocolSyncLoop loop,
  ProtocolQueueAdapter queue,
  AuthoritativeCheckpointTarget target, {
  Duration timeout = const Duration(seconds: 30),
  Duration pollInterval = const Duration(milliseconds: 50),
  CaughtUpCancellationToken? cancellation,
  String? expectedGeneration,
  bool Function()? isOnline,
  DateTime Function()? now,
}) async {
  if (timeout.isNegative || pollInterval.isNegative) {
    throw RangeError('timeout and pollInterval must be non-negative');
  }
  validateAuthoritativeCheckpointTarget(
    target,
    expectedGeneration: expectedGeneration,
  );
  final clock = now ?? DateTime.now;
  final online = isOnline ?? loop.isOnline;
  final startedAt = clock();
  final deadline = startedAt.add(timeout);
  var checkpoint = await queue.pullCheckpoint();
  _assertCheckpoint(checkpoint, local: true);

  if (checkpointReached(checkpoint, target.checkpoint)) {
    return CaughtUpResult(
      targetCheckpoint: target.checkpoint,
      checkpoint: checkpoint,
      generation: target.generation,
      elapsed: clock().difference(startedAt),
      cycles: 0,
      alreadyCaughtUp: true,
    );
  }

  var cycles = 0;
  while (true) {
    if (cancellation?.isCancelled ?? false) {
      throw CaughtUpBarrierException(
        CaughtUpBarrierErrorCode.cancelled,
        'caught-up wait was cancelled',
        targetCheckpoint: target.checkpoint,
        localCheckpoint: checkpoint,
      );
    }
    if (!online() || loop.state.status == ProtocolSyncStatus.offline) {
      throw CaughtUpBarrierException(
        CaughtUpBarrierErrorCode.offline,
        'cannot establish authoritative freshness while offline',
        targetCheckpoint: target.checkpoint,
        localCheckpoint: checkpoint,
      );
    }
    var remaining = deadline.difference(clock());
    if (remaining <= Duration.zero) {
      throw CaughtUpBarrierException(
        CaughtUpBarrierErrorCode.timeout,
        'caught-up wait timed out before the durable checkpoint reached the target',
        targetCheckpoint: target.checkpoint,
        localCheckpoint: checkpoint,
      );
    }

    final before = checkpoint;
    await _boundedForCaller<ProtocolSyncCycleResult>(
      loop.syncNow(),
      remaining: remaining,
      target: target.checkpoint,
      local: checkpoint,
      cancellation: cancellation,
    );
    cycles++;

    // Durable storage is the only freshness evidence; the in-memory cycle
    // result is intentionally ignored here.
    checkpoint = await queue.pullCheckpoint();
    _assertCheckpoint(checkpoint, local: true);
    if (checkpointReached(checkpoint, target.checkpoint)) {
      return CaughtUpResult(
        targetCheckpoint: target.checkpoint,
        checkpoint: checkpoint,
        generation: target.generation,
        elapsed: clock().difference(startedAt),
        cycles: cycles,
        alreadyCaughtUp: false,
      );
    }

    if (checkpoint == before && pollInterval > Duration.zero) {
      remaining = deadline.difference(clock());
      final delay = pollInterval < remaining ? pollInterval : remaining;
      await _boundedForCaller<void>(
        Future<void>.delayed(delay),
        remaining: remaining,
        target: target.checkpoint,
        local: checkpoint,
        cancellation: cancellation,
      );
    }
  }
}

Future<CaughtUpResult> requestAndAwaitCaughtUp(
  ProtocolSyncLoop loop,
  ProtocolQueueAdapter queue,
  AuthoritativeCheckpointRequester requester, {
  Duration timeout = const Duration(seconds: 30),
  Duration pollInterval = const Duration(milliseconds: 50),
  CaughtUpCancellationToken? cancellation,
  String? expectedGeneration,
  bool Function()? isOnline,
  DateTime Function()? now,
}) async {
  if (timeout.isNegative) throw RangeError('timeout must be non-negative');
  final clock = now ?? DateTime.now;
  final startedAt = clock();
  final requestCancellation = CaughtUpCancellationToken();
  StreamSubscription<void>? mirror;
  if (cancellation != null) {
    if (cancellation.isCancelled) {
      requestCancellation.cancel();
    } else {
      // A tiny stream bridge keeps request cancellation independent from the
      // later shared sync cycle.
      mirror = Stream<void>.fromFuture(cancellation.whenCancelled)
          .listen((_) => requestCancellation.cancel());
    }
  }
  try {
    final target = await _boundedForCaller<AuthoritativeCheckpointTarget>(
      requester.requestCheckpoint(requestCancellation),
      remaining: timeout,
      target: '0',
      cancellation: cancellation,
    );
    final elapsed = clock().difference(startedAt);
    final remaining = timeout - elapsed;
    return awaitCaughtUp(
      loop,
      queue,
      target,
      timeout: remaining.isNegative ? Duration.zero : remaining,
      pollInterval: pollInterval,
      cancellation: cancellation,
      expectedGeneration: expectedGeneration,
      isOnline: isOnline,
      now: clock,
    );
  } on CaughtUpBarrierException catch (error) {
    if (error.code == CaughtUpBarrierErrorCode.timeout ||
        error.code == CaughtUpBarrierErrorCode.cancelled) {
      requestCancellation.cancel();
    }
    rethrow;
  } finally {
    await mirror?.cancel();
  }
}

File: clients/ts/src/caught-up.ts
import type {
  ProtocolQueueAdapter,
  ProtocolSyncCycleResult,
  ProtocolSyncState,
} from './sync-loop.js';

const CANONICAL_CHECKPOINT = /^(?:0|[1-9]\d*)$/;

/**
 * An authoritative source position representing all changes committed before
 * the checkpoint request completed.
 *
 * `generation` is optional in protocol-v1 so existing full-replication servers
 * can adopt the barrier without inventing scope semantics. Selective-sync
 * scopes must bind it once scope identity lands (DEN-139).
 */
export interface AuthoritativeCheckpointTarget {
  protocolVersion: 1;
  checkpoint: string;
  generation?: string;
}

/** Optional transport capability; deliberately separate from ProtocolTransport. */
export interface AuthoritativeCheckpointRequester {
  requestCheckpoint(signal: AbortSignal): Promise<AuthoritativeCheckpointTarget>;
}

/** Structural sync-loop surface used by the barrier. */
export interface CaughtUpSyncLoop {
  readonly state?: Readonly<ProtocolSyncState>;
  syncNow(): Promise<ProtocolSyncCycleResult>;
}

export interface AwaitCaughtUpOptions {
  /** Bound the caller's wait. The shared sync cycle itself is never aborted. */
  timeoutMs?: number;
  /** Cancels this waiter only; it never cancels another waiter/background sync. */
  signal?: AbortSignal;
  /** Poll delay when a successful cycle made no checkpoint progress. */
  pollIntervalMs?: number;
  /** Override connectivity detection for non-browser hosts/tests. */
  isOnline?: () => boolean;
  /** Expected checkpoint namespace/reset/scope generation, when one exists. */
  expectedGeneration?: string;
  /** Injectable monotonic-ish wall clock for elapsed-time reporting/tests. */
  now?: () => number;
}

export type CaughtUpBarrierErrorCode =
  | 'CAUGHT_UP_INVALID_TARGET'
  | 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT'
  | 'CAUGHT_UP_INVALIDATED'
  | 'CAUGHT_UP_TIMEOUT'
  | 'CAUGHT_UP_CANCELLED'
  | 'CAUGHT_UP_OFFLINE';

export class CaughtUpBarrierError extends Error {
  constructor(
    public readonly code: CaughtUpBarrierErrorCode,
    message: string,
    public readonly targetCheckpoint?: string,
    public readonly localCheckpoint?: string,
  ) {
    super(message);
    this.name = 'CaughtUpBarrierError';
  }
}

export interface CaughtUpResult {
  targetCheckpoint: string;
  checkpoint: string;
  generation?: string;
  elapsedMs: number;
  cycles: number;
  alreadyCaughtUp: boolean;
}

function assertCheckpoint(
  checkpoint: string,
  code: 'CAUGHT_UP_INVALID_TARGET' | 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT',
): void {
  if (!CANONICAL_CHECKPOINT.test(checkpoint)) {
    throw new CaughtUpBarrierError(
      code,
      `${code === 'CAUGHT_UP_INVALID_TARGET' ? 'target' : 'local'} checkpoint must be a canonical unsigned decimal string`,
      code === 'CAUGHT_UP_INVALID_TARGET' ? checkpoint : undefined,
      code === 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT' ? checkpoint : undefined,
    );
  }
}

export function checkpointReached(local: string, target: string): boolean {
  assertCheckpoint(local, 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT');
  assertCheckpoint(target, 'CAUGHT_UP_INVALID_TARGET');
  return BigInt(local) >= BigInt(target);
}

export function validateAuthoritativeCheckpointTarget(
  target: AuthoritativeCheckpointTarget,
  expectedGeneration?: string,
): void {
  if (target.protocolVersion !== 1) {
    throw new CaughtUpBarrierError(
      'CAUGHT_UP_INVALID_TARGET',
      'authoritative checkpoint target has unsupported protocolVersion',
      target.checkpoint,
    );
  }
  assertCheckpoint(target.checkpoint, 'CAUGHT_UP_INVALID_TARGET');
  if (target.generation !== undefined && target.generation.length === 0) {
    throw new CaughtUpBarrierError(
      'CAUGHT_UP_INVALID_TARGET',
      'authoritative checkpoint generation must be non-empty when present',
      target.checkpoint,
    );
  }
  if (
    expectedGeneration !== undefined &&
    target.generation !== expectedGeneration
  ) {
    throw new CaughtUpBarrierError(
      'CAUGHT_UP_INVALIDATED',
      `checkpoint generation ${JSON.stringify(target.generation ?? null)} does not match expected generation ${JSON.stringify(expectedGeneration)}`,
      target.checkpoint,
    );
  }
}

function finiteNonNegative(value: number, name: string): number {
  if (!Number.isFinite(value) || value < 0) {
    throw new RangeError(`${name} must be a finite non-negative number`);
  }
  return value;
}

function callerCancelled(target: string, local?: string): CaughtUpBarrierError {
  return new CaughtUpBarrierError(
    'CAUGHT_UP_CANCELLED',
    'caught-up wait was cancelled',
    target,
    local,
  );
}

function callerTimedOut(target: string, local?: string): CaughtUpBarrierError {
  return new CaughtUpBarrierError(
    'CAUGHT_UP_TIMEOUT',
    'caught-up wait timed out before the durable checkpoint reached the target',
    target,
    local,
  );
}

/**
 * Wait for one promise while bounding only this caller.
 *
 * Deliberately does not abort `promise`: when it is ProtocolSyncLoop.syncNow(),
 * that promise may be shared by background work and other caught-up waiters.
 */
function waitForCaller<T>(
  promise: Promise<T>,
  signal: AbortSignal | undefined,
  timeoutMs: number,
  target: string,
  local?: string,
): Promise<T> {
  if (signal?.aborted) return Promise.reject(callerCancelled(target, local));
  if (timeoutMs <= 0) return Promise.reject(callerTimedOut(target, local));

  return new Promise<T>((resolve, reject) => {
    let settled = false;
    const finish = (fn: () => void) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      signal?.removeEventListener('abort', onAbort);
      fn();
    };
    const onAbort = () =>
      finish(() => reject(callerCancelled(target, local)));
    const timer = setTimeout(
      () => finish(() => reject(callerTimedOut(target, local))),
      timeoutMs,
    );
    signal?.addEventListener('abort', onAbort, { once: true });
    promise.then(
      (value) => finish(() => resolve(value)),
      (error) => finish(() => reject(error)),
    );
  });
}

function delayForCaller(
  delayMs: number,
  signal: AbortSignal | undefined,
  timeoutMs: number,
  target: string,
  local: string,
): Promise<void> {
  return waitForCaller(
    new Promise<void>((resolve) => setTimeout(resolve, delayMs)),
    signal,
    timeoutMs,
    target,
    local,
  );
}

/**
 * Drive the ordinary single-flight sync loop until its *durable* checkpoint is
 * at or beyond `target.checkpoint`.
 *
 * `idle`, a WebSocket wake, or a successful network response is never treated
 * as freshness evidence. Completion is decided only by rereading
 * `queue.pullCheckpoint()` after synchronization.
 */
export async function awaitCaughtUp(
  loop: CaughtUpSyncLoop,
  queue: Pick<ProtocolQueueAdapter, 'pullCheckpoint'>,
  target: AuthoritativeCheckpointTarget,
  options: AwaitCaughtUpOptions = {},
): Promise<CaughtUpResult> {
  validateAuthoritativeCheckpointTarget(target, options.expectedGeneration);
  const now = options.now ?? Date.now;
  const timeoutMs = finiteNonNegative(options.timeoutMs ?? 30_000, 'timeoutMs');
  const pollIntervalMs = finiteNonNegative(
    options.pollIntervalMs ?? 50,
    'pollIntervalMs',
  );
  const isOnline =
    options.isOnline ??
    (() => typeof navigator === 'undefined' || navigator.onLine !== false);
  const startedAt = now();
  const deadline = startedAt + timeoutMs;
  let checkpoint = await queue.pullCheckpoint();
  assertCheckpoint(checkpoint, 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT');

  if (checkpointReached(checkpoint, target.checkpoint)) {
    return {
      targetCheckpoint: target.checkpoint,
      checkpoint,
      generation: target.generation,
      elapsedMs: Math.max(0, now() - startedAt),
      cycles: 0,
      alreadyCaughtUp: true,
    };
  }

  let cycles = 0;
  for (;;) {
    if (options.signal?.aborted) {
      throw callerCancelled(target.checkpoint, checkpoint);
    }
    if (!isOnline() || loop.state?.status === 'offline') {
      throw new CaughtUpBarrierError(
        'CAUGHT_UP_OFFLINE',
        'cannot establish authoritative freshness while offline',
        target.checkpoint,
        checkpoint,
      );
    }
    const remaining = deadline - now();
    if (remaining <= 0) throw callerTimedOut(target.checkpoint, checkpoint);

    const before = checkpoint;
    await waitForCaller(
      loop.syncNow(),
      options.signal,
      remaining,
      target.checkpoint,
      checkpoint,
    );
    cycles += 1;

    // The durable store is the evidence. Never trust the cycle's in-memory
    // result checkpoint here: an application adapter may have failed to persist.
    checkpoint = await queue.pullCheckpoint();
    assertCheckpoint(checkpoint, 'CAUGHT_UP_INVALID_LOCAL_CHECKPOINT');
    if (checkpointReached(checkpoint, target.checkpoint)) {
      return {
        targetCheckpoint: target.checkpoint,
        checkpoint,
        generation: target.generation,
        elapsedMs: Math.max(0, now() - startedAt),
        cycles,
        alreadyCaughtUp: false,
      };
    }

    if (checkpoint === before && pollIntervalMs > 0) {
      const afterCycleRemaining = deadline - now();
      if (afterCycleRemaining <= 0) {
        throw callerTimedOut(target.checkpoint, checkpoint);
      }
      await delayForCaller(
        Math.min(pollIntervalMs, afterCycleRemaining),
        options.signal,
        afterCycleRemaining,
        target.checkpoint,
        checkpoint,
      );
    }
  }
}

/**
 * Ask the server for a source position representing "now", then wait until the
 * durable local checkpoint reaches it. Timeout covers both phases.
 */
export async function requestAndAwaitCaughtUp(
  loop: CaughtUpSyncLoop,
  queue: Pick<ProtocolQueueAdapter, 'pullCheckpoint'>,
  requester: AuthoritativeCheckpointRequester,
  options: AwaitCaughtUpOptions = {},
): Promise<CaughtUpResult> {
  const now = options.now ?? Date.now;
  const timeoutMs = finiteNonNegative(options.timeoutMs ?? 30_000, 'timeoutMs');
  const startedAt = now();
  const requestController = new AbortController();
  const onAbort = () => requestController.abort();
  options.signal?.addEventListener('abort', onAbort, { once: true });
  try {
    const target = await waitForCaller(
      requester.requestCheckpoint(requestController.signal),
      options.signal,
      timeoutMs,
      '0',
    ).catch((error) => {
      if (
        error instanceof CaughtUpBarrierError &&
        (error.code === 'CAUGHT_UP_TIMEOUT' ||
          error.code === 'CAUGHT_UP_CANCELLED')
      ) {
        requestController.abort();
      }
      throw error;
    });
    const elapsed = Math.max(0, now() - startedAt);
    return awaitCaughtUp(loop, queue, target, {
      ...options,
      timeoutMs: Math.max(0, timeoutMs - elapsed),
      now,
    });
  } finally {
    options.signal?.removeEventListener('abort', onAbort);
  }
}
